src/solver/conv_asm_implicit_gemm_gtc_bwd_nhwc.cpp

```diff
@@ -777,6 +777,9 @@ bool PerformanceConfigAsmImplicitGemmGTCBwdXdlopsNHWC::IsValid(const Convolution
         if(ctx.IsFp16() && gemm_k_global_split != 0 && vector_store != 1)
             return false;
 
+    if(ctx.conv_problem.GetConv().attribute.deterministic.Get() && gemm_k_global_split != 0)
```

Comment:
Please explain how it is expected to provide determinism.

Reply:
when `gemm_k_global_split` is not zero, it indicate the kernel will split gemm_k, and finally use atomic add to accumulate the result.

Reply:
Thanks, but this was already clear. The question is how this check in `TuningConfig::IsValid()` ensures determinism. The answer is non-trivial, so I had to investigate the code myself:

- If the ProblemConfig is tuned, then
  - This check ensures that TuningConfig is refused when it requires non-zero `gemm_k_global_split`. Then the common machinery leverages `HeuristicInit()` to obtain the "default" TuningConfig, which shouldn't, presumably, require "atomic add" (the default TuningConfig should contain `gemm_k_global_split == 0`).
- Otherwise, `HeuristicInit()` is directly used to obtain the "default" TuningConfig.
- `HeuristicInit()` calls `IsValid()` to make sure that the result of the `for(const auto& config : GetBwdXdlopsNHWCConfigList())` loop contains `gemm_k_global_split == 0`.

However when all attempts in the loop fail (of when `((m_per_block == 0 && n_per_block == 0 && k_per_block == 0) || not_support_vector_store)` is true), `HeuristicInit()` uses the `find_with_gemm_k_pad` lambda. The lambda uses `kernel_param_list` to initialize the result. `kernel_param_list` contains non-zero initializers for `gemm_k_global_split`, and the `find_with_gemm_k_pad` lambda does not check this.

As a result, non-zero `gemm_k_global_split` can be produced.

Reply:
Please double-check the issue and provide a fix if necessary.

Please also add comment (to the source code) that explains the machinery, i.e. how the check ensures determinism.

Reply:
@atamazov I think I understand your question here. Actually, there is an internal rule when we create the `kernel_param_list`. There always will be a config with `gemm_k_global_split=0`  in front of a almost the same config, with only `gemm_k_global_split=1`, so it will never choose the *=1 one. And indeed in this `HeuristicInit()` we never wish to find a config with `gemm_k_global_split=1`.
But yes, I think this need clarification or comments, or a more robust check inside the `HeuristicInit()`.